providers/push-api/src/lib/push-api.provider.ts

```diff
@@ -0,0 +1,112 @@
+import {
+  ChannelTypeEnum,
+  ISendMessageSuccessResponse,
+  IPushOptions,
+  IPushProvider,
+} from '@novu/stateless';
+import webpush from 'web-push';
+import crypto from 'crypto';
+
+class ErrorBase extends Error {
+  constructor(message?: string) {
+    super(message);
+
+    this.name = this.constructor.name;
+    Error.captureStackTrace(this, this.constructor);
+  }
+}
+
+class PushError extends ErrorBase {
+  constructor(public readonly failures: webpush.SendResult[], message: string) {
+    super();
+    Object.setPrototypeOf(this, new.target.prototype);
+    this.failures = failures;
+    this.message =
+      message +
+      '\nFailures:\n' +
+      failures.map((failure) => failure.statusCode).join('\n');
+  }
+}
+
+export class PushApiPushProvider implements IPushProvider {
+  channelType = ChannelTypeEnum.PUSH as ChannelTypeEnum.PUSH;
+  id: 'push-api';
+
+  constructor(
+    private config: {
+      vapidPrivateKey: string;
+      vapidPublicKey: string;
+    }
+  ) {}
+
+  async sendPush(
+    title: string,
+    content: string,
+    subscription: webpush.PushSubscription
+  ) {
+    const pushOptions: webpush.RequestOptions = {
+      vapidDetails: {
+        subject: 'mailto:support@novu.co',
+        privateKey: this.config.vapidPrivateKey,
+        publicKey: this.config.vapidPublicKey,
+      },
+    };
+
+    return await webpush.sendNotification(
+      subscription,
+      JSON.stringify({
+        title,
+        body: content,
+      }),
+      pushOptions
+    );
+  }
+
+  getId = () => crypto.randomBytes(20).toString('hex');
+
+  async sendMessage(
+    options: IPushOptions
+  ): Promise<ISendMessageSuccessResponse> {
+    const subscriptions = options.target;
+
+    const failures: webpush.SendResult[] = [];
+    const id = this.getId();
+
+    for (const subscriptionString of subscriptions) {
```

Comment:
Why do we have their list of subscriptions? on this level, we should have the Novu subscriber as a target and it should be a single recipient.

Reply:
So with web push, you can have multiple devices which all need to receive a push. 

How I understood the targets, each target represented a push end point for a single subscriber. So in the case of mobile push through APS, they could have multiple iOS devices, but they are logged in with the same credential (subscriber id). 

In this case, I have a desktop browser and an PWA browser running on my phone. Both would be subscribed with the same user.

Reply:
The other thing to note, is web push requires an endpoint and key information. So rather than a single endpoint/target string, the endpoint is really an encoded string with an endpoint and the key signatures.